onnx/onnx-ml.proto

```diff
@@ -199,6 +207,112 @@ message NodeProto {
+//      nodes in TrainingInfoProto.algorithm.nodes and
+//      TrainingInfoProto.initialization.nodes. It implies that training
+//      nodes cannot produce values with names identical to inference
+//      initializers.
+message TrainingInfoProto {
+  // This field describes a graph to compute the initial tensors
+  // upon starting the training process. Initialization graph has no input
+  // and can have multiple outputs. Usually, trainable tensors in neural
+  // networks are randomly initialized. To achieve that, for each tensor,
+  // the user can put a random operator such as RandomNormal or
+  // RandomUniform in TrainingInfoProto.initialization.node and assign its
+  // random output to the specific tensor using "initialization_binding".
+  // This graph can also set the initializers in "algorithm"; a use case is
+  // resetting the number of training iteration to zero.
+  //
+  // By default, this field is an empty graph and its evaluation does not
+  // produce any output.
+  optional GraphProto initialization = 1;
+
+  // This field represents a training algorithm step. Given required inputs,
+  // it computes outputs to update initializers in its own initializer
+  // list and ModelProto.graph.initializer. In general, this graph contains
+  // loss node, gradient node, optimizer node, increment of iteration count,
+  // and calls to the inference graph.
+  //
+  // The field algorithm.node can invoke ModelProto.graph as a node.
+  // The initializers defined in ModelProto.graph.initializer are visible in
+  // the "algorithm".
+  //
+  // The field MUST be present.
+  optional GraphProto algorithm = 2;
+
+  // This field specifies the bindings from the outputs of "initialization" to
+  // the initializers in "algorithm" and "ModelProto.graph". See
+  // "update_binding" below for its format.
+  optional GraphProto initialization_binding = 3;
+
+  // Gradient-based training is usually an iterative procedure. In one gradient
+  // descent iteration, we apply
+  //
+  // x = x - r * g
+  //
+  // where "x" is the optimized tensor, "r" stands for learning rate, and "g" is
+  // gradient of "x" with respect to a chosen loss. To avoid adding assignments
+  // into the training graph, we split the update equation into
+  //
+  // y = x - r * g
+  // x = y
+  //
+  // The user needs to save "y = x - r * g" into TrainingInfoProto.algorithm. To
+  // tell that "y" should be assigned to "x", the field "update_binding" may
+  // contain a pair of strings, "x" (key of StringStringEntryProto) and
+  // "y" (value of StringStringEntryProto).
+  // For a neural network with multiple trainable tensors, there can
+  // be multiple key-value pairs in "update_binding".
+  //
+  // One initializer can only be assigned up to once; it implies that the
+  // keys in "update_binding" must be unique.
+  //
+  // Whether an initializer is trainable is determined by the keys of
+  // key-value pairs in "update_binding". If the name of an initializer
```

Comment:
Trainable weights must be an input in `ModelProto.graph.input`?

Reply:
I agree with you @wschin :+1:

Reply:
I think they are two separate aspects. (a) Should we pass the trainable-weights as explicit-inputs to ModelProto.graph or should we treat trainable-weights as "shared variables" with initializers of the same name in ModelProto.graph? (b) How does a runtime figure out the set of trainable inputs? You are right with respect to (b): it is sufficient to look at only update-bindings. However, this does not depend on the answer we choose for (a). I think for (a) treating the trainable-weights as explicit-inputs is more consistent with the existing spec (it will surprise fewer people/implementations who may be assuming this), and it seems cleaner.

Reply:
I am not sure if I fully understand the question (a). I think for (a), the current training spec is doing `treat trainable-weights as "shared variables" with initializers of the same name in ModelProto.graph`, because `ModelProto.graph.initializer` is visible to `TrainingInfoProto.algorithm`.

I feel there should some kinds of variables:
1. trainable, inference-time constant: in `ModelProto.graph.initializer` and `update_binding`
2. trainable, not inference-time constant: in `ModelProto.graph.input`, `ModelProto.graph.initializer`, and `update_binding`
3. not trainable, inference-time constant: in `ModelProto.graph.initializer`
4. not trainable, not inference-time constant:  in `ModelProto.graph.input`, `ModelProto.graph.initializer`

If we always put `trainable` in `ModelProto.graph.input`, the categories (1) and (2) will be merged into one category. Is it ok?

Reply:
I personally feel it is okay to merge categories (1) and (2) into one. But it depends on the overall scenarios we want to support: do we imagine that after training, we would export an inference-model optimized for inference, or do we always want to leave the model in the trainable-format forever even when used primarily for inference? In the first scenario, the distinction is not relevant. In the second scenario, the distinction might be slightly relevant.

Reply:
I feel ONNX can have a tool to convert and optimize trainable models to inference-only models. For inference-focusing models, I feel there should be no training information for smaller model size. With the optimizer states in Adam, Adagrad, and so on, the model size can be 3x larger than the inference-only version.

Reply:
I completely agree that training models should be able to be optimized for deployment, but if you combine (1) and (2) then you can't _mechanically_ infer which trainable inputs/initializers should be inference-time constants.

Reply:
I think there is an answer to @TMVector 's concern. Well, two answers: (a) We can use the inputs of the trainable-graph (the update algorithm) as the inputs of the optimized inference-model generated after training, (b) The optimization-tool can give users control over the generated model to control whether trained-parameters should be made inputs of the generated model.

Reply:
Here is my mental model of what's in the model-file:
```
   function inference(input X, input W) {
      initial value W = ..., C = ...;
   }
   function training-step(input X, input Y) {
      static TW; // trainable weight
      ...
      Z = inference(X, TW)
      ...
      TW = update(...)
   }
```
If we want a trainable weight that should not be a constant in an inference-model that is exported, we can make this an input of the update algorithm (option 1) or just make it user-controllable for the optimizer-and-export step.

I am not saying this is ideal. It is a compromise either way. I think this will be less disruptive. The problem is that the earlier design of the inference-model had some partial-support built-in in anticipation of training, and now that we have full training-support we can simplify things, but it will have an impact on existing infrastructure.

Reply:
(a) might not always work. Let's consider a real click-prediction problem. The user saved a trainable linear model (trainable parameters: linear coefficient vector `w` and a bias value `b`) and invoked a runtime to train it. 
Let's then assume
- `w` and `b` are in `ModelProto.graph.input` and `ModelProto.graph.initializer`.
- `w` and `b` are in `TrainingInfoProto.algorithm.input` and `ModelProto.graph.initializer`.
- `w` and `b` can be found as keys in `TrainingInfoProto.update_binding`
In the inference-only model, would (a) make `b` an input? Note that `b` must be an input and `w` should not.

For (b), the user who wants to use a trained model might not be the author of the original trainable model. It's very difficult (and almost impossible when you have large neural networks) for the end user to figure out the mathematical meanings of those inputs and initializers.

Reply:
@gramalingam, currently we don't have any trainable graph exported. I feel switching to a simpler semantic might lead to a easier life.

@spandantiwari, how troublesome it is to Pytorch exporter if we only use `update_binding` to determine trainable tensors?

Reply:
PyTorch exporter does not have any training semantics at all today. But this question is quite relevant going forward.

The two choices:
1) Specify trainable parameters only through `update_binding` mapping.
2) In addition to 1), the trainable parameters must also be listed as forward graph input.

I find that just having 1) is necessary and sufficient. I see two advantages:
a) The spec is clear and minimal. 
b) It decouples the semantics of training vs inference from existing spec (and consequently existing exporters) to just the newly added training spec. So, if someone wants to encapsulate a graph exported for inference only in `TrainingInfoProto.algorithm`, they don't have to go back and re-export the graph to satisfy 2), i.e. re-export graph with trainable parameters listed as graph input. They can just take the existing graph as is and encode the differentiation between trainable and non-trainable parameters in `update_binding`. 

For keeping 2), I can see that some backends already may have optimizations that are based on whether an initializer is part of the graph input or not. For that, we can be flexible and choose to go with 2) even though it is not necessary. 

Finally, even if we adopt only 1) in the spec, it is quite possible that some exporters choose to export training graphs that have trainable parameters listed as graph input, and that should be fine. In fact, this doesn't even have to be restricted to training export only, but can be done for inference graphs also. And the backends can still keep their logic for which parameters to optimize.